py/obj.c

```diff
@@ -159,28 +159,25 @@ bool mp_obj_is_callable(mp_obj_t o_in) {
 // Furthermore, from the v3.4.2 code for object.c: "Practical amendments: If rich
 // comparison returns NotImplemented, == and != are decided by comparing the object
 // pointer."
-bool mp_obj_equal(mp_obj_t o1, mp_obj_t o2) {
-    // Float (and complex) NaN is never equal to anything, not even itself,
-    // so we must have a special check here to cover those cases.
-    if (o1 == o2
-        #if MICROPY_PY_BUILTINS_FLOAT
-        && !mp_obj_is_float(o1)
-        #endif
-        #if MICROPY_PY_BUILTINS_COMPLEX
-        && !mp_obj_is_type(o1, &mp_type_complex)
-        #endif
-        ) {
-        return true;
+mp_obj_t mp_obj_equal_bop(mp_obj_t o1, mp_obj_t o2, bool not_equal) {
+    mp_obj_t local_true = not_equal ? mp_const_false : mp_const_true;
+    mp_obj_t local_false = not_equal ? mp_const_true : mp_const_false;
+
+    // Shortcut for very common cases
+    if (o1 == o2 &&
+        (mp_obj_is_small_int(o1) || o1 == mp_const_none) ) {
+        return local_true;
     }
+
     if (o1 == mp_const_none || o2 == mp_const_none) {
```

Comment:
I think this if statement needs to be removed (or changed somehow...) because it's up to a user type if it's equal to None or not.

Reply:
Quite right. It *should* just work if I take this out.

Reply:
I check, and everything still works when simply removing this path.